### PR TITLE
RSVP Projects for Events

### PR DESCRIPTION
**Pull request by marlonkeating**
Initial implementation of Projects RSVP-ing for events.

### CODE REVIEW

**Comment by PeterBreen** [comment]
Mobile and desktop are in MVP status, will work on testing.

Known changes from the design given (wontfix for this release, possibly after):
 - drop shadow deleted
 - (desktop) logo image does not superimpose on blue banner box; we can't rely on users always providing both square logo images and high-contrast to the blue/white background images, so that must be revisited.
 - (desktop) list of event details in the top box is always 1 column; we don't need 2-col to save vertical height because the logo is taking up more vertical space, so I reused our existing 1-col design (see a project profile page) for consistency
   - one of the icons isn't right, it asked for FA premium and we still load FA free (calendar-day) so I made a sub
 
 
Issues to fix:
 - routing/URL structure (possibly also search engine de-emphasis of these pages?)
 - all the testing 
 - comment out or enable placeholder (PH) button
 - enable and connect toasts to RSVP/cancel/etc actions

File: common/components/common/event_projects/AboutProjectEventDisplay.jsx [hunk 1]
```diff
@@ -0,0 +1,439 @@
...
+          )}
+
+          <h3>Planned Scope</h3>
+          {eventProject?.event_project_scope ? (
+            <p>{eventProject.event_project_scope}</p>
+          ) : (
+            comingSoonMsg
+          )}
+
+          {/*TODO: Show newlines*/}
```
Comment (PeterBreen) [inline common/components/common/event_projects/AboutProjectEventDisplay.jsx:219]: todo, but I think this one's done? If I remember the tests we ran before deploy....?

File: common/components/common/event_projects/AboutProjectEventDisplay.jsx [hunk 2]
```diff
@@ -0,0 +1,439 @@
...
+
+  _renderTeam(): React$Node {
+    const eventProject: EventProjectAPIDetails = this.state.eventProject;
+    const numVolunteers = eventProject["project_volunteers"].length || 0;
+    const groupedVolunteers = _.groupBy(
+      eventProject.project_volunteers,
+      "roleTag.subcategory"
+    );
+    const sortedVolunteers = Object.entries(groupedVolunteers).sort();
+    // may not need all these consts; test what happens when project_volunteers is null/empty
```
Comment (PeterBreen) [inline common/components/common/event_projects/AboutProjectEventDisplay.jsx:309]: a todo but not labeled as such; I don't remember testing this

File: common/components/common/event_projects/AboutProjectEventDisplay.jsx [hunk 3]
```diff
@@ -0,0 +1,439 @@
...
+
+    // end result: group and count volunteers by subcategory, then render VolunteerSection for each subcategory
+    // TODO: Do we need the additional props used on AboutProject for VolunteerSection?
```
Comment (PeterBreen) [inline common/components/common/event_projects/AboutProjectEventDisplay.jsx:312]: todo but I think the answer is 'no' and this section's getting worked on anyway

File: common/components/componentsBySection/CreateEvent/AboutEventDisplay.jsx [hunk 4]
```diff
@@ -179,12 +200,67 @@ class AboutEventDisplay extends React.PureComponent<Props, State> {
...
         href={url}
+        target="_blank"
       >
         {text}
       </Button>
     );
   }
 
+  _renderRSVPAsProjectOwnerButton(): ?$React$Node {
+    // TODO: Don't show when a user has rsvp-ed all their projects with this event
```
Comment (PeterBreen) [inline common/components/componentsBySection/CreateEvent/AboutEventDisplay.jsx:211]: todo, conditional button render

File: common/components/componentsBySection/CreateEventProject/CreateEventProjectScope.jsx [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+
+        <TextFormField
+          id="schedule"
+          label="Schedule for Hackathon Team (please include a time zone)"
+          type={TextFormFieldType.MultiLine}
+          rows={6}
+          placeholder={
+            "Please provide a schedule to guide your team through the hackathon.\n\n" +
+            'e.g. 11:00 AM PST: Teams start working on the "Climathonians" Mobile App.\n\n' +
+            "12:00 AM PST: Break"
```
Comment (PeterBreen) [inline common/components/componentsBySection/CreateEventProject/CreateEventProjectScope.jsx:98]: Not a todo item for dev, but if this function gets further design review as we fix bugs and add features, I think we should make a note of this. The instant someone types a character this entire helper text disappears -- it's not best practice to put so much in a placeholder.

File: common/components/componentsBySection/CreateEventProject/CreateEventProjectSelect.jsx [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+type State = {|
+  owned_projects: $ReadOnlyArray<MyProjectData>,
+  isAlreadySelected: boolean,
+  formIsValid: boolean,
+  validations: $ReadOnlyArray<Validator>,
+|} & FormStateBase<FormFields>;
+
+/**
+ * Encapsulates form for selecting project to RSVP for event
+ */
```
Comment (PeterBreen) [inline common/components/componentsBySection/CreateEventProject/CreateEventProjectSelect.jsx:41]: I've seen a lot of these and I'm just commenting to say that I like seeing them. A little inline help on what a component does is excellent practice. 👍

File: common/components/controllers/CreateEventProjectController.jsx [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+  updatePageUrl() {
+    if (this.state.projectId && !url.argument("project_id")) {
+      url.updateArgs({ project_id: this.state.projectId });
+    }
+    utils.navigateToTopOfPage();
+  }
+
+  loadEventProjectDetails(eventProject: EventProjectAPIDetails): void {
+    if (!CurrentUser.isCoOwnerOrOwner(eventProject) && !CurrentUser.isStaff()) {
+      // TODO: Handle someone other than owner
```
Comment (PeterBreen) [inline common/components/controllers/CreateEventProjectController.jsx:128]: todo on handling loads for non-owners?

File: common/components/controllers/CreateEventProjectController.jsx [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+    event: SyntheticEvent<HTMLFormElement>,
+    formRef: HTMLFormElement,
+    onSubmitSuccess: (EventProjectAPIDetails, () => void) => void
+  ): void {
+    const formSubmitUrl: string = `/api/event/${this.state.eventId}/projects/${this.state.projectId}/create/`;
+    api.postForm(
+      formSubmitUrl,
+      formRef,
+      onSubmitSuccess,
+      response => null /* TODO: Report error to user */
```
Comment (PeterBreen) [inline common/components/controllers/CreateEventProjectController.jsx:156]: todo, probably going to require toast implementation to put the error on the frontend the way we want?

File: common/components/controllers/CreateProjectController.jsx [hunk 9]
```diff
@@ -176,11 +177,19 @@ class CreateProjectController extends React.PureComponent<{||}, State> {
     });
     metrics.logProjectCreated(CurrentUser.userID());
     // TODO: Fix bug with switching to this section without page reload
```
Comment (PeterBreen) [inline common/components/controllers/CreateProjectController.jsx:179]: existing todo item, not part of this PR

File: common/components/forms/FormWorkflow.jsx [hunk 10]
```diff
@@ -105,6 +119,7 @@ class FormWorkflow<T> extends React.PureComponent<Props<T>, State<T>> {
     });
   }
 
+  // TODO: Still necessary?
```
Comment (PeterBreen) [inline common/components/forms/FormWorkflow.jsx:122]: is it? 😄 

see also L133, L149, same questions

File: common/components/forms/ImageCropUploadFormElement.jsx [hunk 11]
```diff
@@ -37,27 +40,41 @@ class ImageCropUploadFormElement extends React.PureComponent<Props, State> {
...
     this.setState(
       { currentImage: fileInfo },
       () => this.props.onUpdate && this.props.onUpdate(fileInfo)
     );
+    UniversalDispatcher.dispatch({
+      type: "UPDATE_FORM_FIELD",
+      fieldName: this.props.form_id,
+      fieldValue: fileInfo,
+    });
+    // TODO: Need this?
```
Comment (PeterBreen) [inline common/components/forms/ImageCropUploadFormElement.jsx:64]: todo item

File: common/components/forms/ImageCropUploadFormElement.jsx [hunk 12]
```diff
@@ -37,27 +40,41 @@ class ImageCropUploadFormElement extends React.PureComponent<Props, State> {
...
     return (
       <div>
         <ImageCropUploadButton
-          currentImage={previewImage}
+          currentImage={this.state.currentImage}
           aspect={this.props.aspect}
           buttonText={this.props.buttonText || "Upload Image"}
           onFileUpload={this._handleFileSelection.bind(this)}
         />
+        {/*TODO: Use HiddenFormField component*/}
```
Comment (PeterBreen) [inline common/components/forms/ImageCropUploadFormElement.jsx:77]: refactor to use the HiddenFormField component

**Review by PeterBreen — CHANGES_REQUESTED** [review]
Okay, I think I listed every todo item that's relevant to this PR. there may be a couple pre-existing ones I missed.

by and large it looks fine, it also felt like I was re-reviewing a lot of code from #802 but that's alright. 

I mentioned it in a comment but I'd also like to say I really appreciate a lot of the comments near the head of a given component file explaining what that component does. 👍  for sure.

Once you've confirmed all the todo items are known issues in trello this is good to go unless you want to handle some outstanding bugs in this branch. Fine either way.

File: civictechprojects/helpers/context_preload.py [hunk 13]
```diff
@@ -7,6 +7,7 @@
 from common.helpers.redirectors import RedirectTo
 from common.helpers.request_helpers import url_params
 
+#TODO: event_project_preload
```
[the remark this answers is not in the document]
Comment (marlonkeating) [reply civictechprojects/helpers/context_preload.py:10]: Listed as "Need headers for Event Project Profile/Project RSVP pages"

File: civictechprojects/forms.py [hunk 14]
```diff
@@ -287,3 +293,51 @@ def delete_group(request, group_id):
...
+            recache_linked = True
+
+        fields_changed |= read_form_field_string(event_project, form, 'goal')
+        fields_changed |= read_form_field_string(event_project, form, 'scope')
+        fields_changed |= read_form_field_string(event_project, form, 'schedule')
+        fields_changed |= read_form_field_string(event_project, form, 'onboarding_notes')
+
+        # TODO: Add to model
+        # if not request.user.is_staff:
+        #     project.project_date_modified = timezone.now()
```
[the remark this answers is not in the document]
Comment (marlonkeating) [reply civictechprojects/forms.py:328]: Listed as "Event Profile: Position project search pulls from main project not event project"

File: civictechprojects/migrations/data_migrations/migrate_event_projects.py [hunk 15]
```diff
@@ -0,0 +1,30 @@
...
+
+
+def get_event_legacy_projects(event):
+    legacy_slugs = event.event_legacy_organization.slugs()
+    if legacy_slugs:
+        return Project.objects.filter(project_organization__name__in=legacy_slugs, is_searchable=True)
+
+
+def migrate_event_projects(*args):
+    # TODO: Migrate existing video events to link_video
```
[the remark this answers is not in the document]
Comment (marlonkeating) [reply civictechprojects/migrations/data_migrations/migrate_event_projects.py:11]: Listed as "Legacy project video links are not playable"

File: civictechprojects/static/css/partials/_Profile.scss [hunk 16]
```diff
@@ -22,6 +22,7 @@
   }
   order: 1;
   flex-basis: 35%;
+  position: relative;
```
[the remark this answers is not in the document]
Comment (marlonkeating) [reply civictechprojects/static/css/partials/_Profile.scss:25]: Yes, it's used to place the play button correctly in the center of the image

File: common/components/controllers/CreateEventProjectController.jsx (hunk 7 above)
Comment (marlonkeating) [reply common/components/controllers/CreateEventProjectController.jsx:128]: Recorded as "Show error for people who are not the project owner or admin"

File: common/components/controllers/CreateEventProjectController.jsx (hunk 8 above)
Comment (marlonkeating) [reply common/components/controllers/CreateEventProjectController.jsx:156]: Recorded as "Display api errors on frontend"

File: common/components/forms/ImageCropUploadFormElement.jsx (hunk 12 above)
Comment (marlonkeating) [reply common/components/forms/ImageCropUploadFormElement.jsx:77]: Added to list
